nominal/nominal.py

```diff
@@ -94,38 +102,64 @@ def upload_tdms(
         channels_to_export: dict[str, TdmsChannel] = {}
         group: TdmsGroup
+        df = None
+
         for group in tdms_file.groups():
+            time_channel = None
             channel: TdmsChannel
+
+            # pull out timestamp column if expected
+            if timestamp_column:
+                for channel in group.channels():
+                    if channel.name == timestamp_column:
+                        time_channel = channel
+                if not time_channel:
+                    logger.info("Skipping channel group \"%s\" because expected timestamp_column \"%s\" does not exist",
+                                channel.group_name, timestamp_column)
+                    continue
+
+            # select channels for export
             for channel in group.channels():
-                # some channels may not have the required properties to construct a time track
-                if ("wf_increment" in channel.properties) and ("wf_start_time" in channel.properties):
+                # do not export timestamp column
+                if channel.name == timestamp_column:
+                    continue
+
+                # export IFF timestamp_column provided OR waveform time properties are set
+                if timestamp_column or (("wf_increment" in channel.properties) and ("wf_start_time" in
+                                                                                    channel.properties)):
+                    # skip if unexpected column length
+                    if timestamp_column and len(channel) != len(time_channel):
+                        logger.info("Skipping channel \"%s\" because length does not match \"%s\"",
+                                    channel.name, timestamp_column)
+                        continue
                     channel_name = f"{channel.group_name.replace(' ', '_')}.{channel.name.replace(' ', '_')}"
                     channels_to_export[channel_name] = channel
 
-        df = pd.DataFrame.from_dict(
-            {
-                channel_name: pd.Series(
-                    data=channel.read_data(), index=channel.time_track(absolute_time=True, accuracy="ns")
-                )
-                for channel_name, channel in channels_to_export.items()
-            }
-        )
+            group_df = pd.DataFrame.from_dict(
+                {
+                    channel_name: pd.Series(
+                        data=channel.read_data(), index=time_channel.read_data() if timestamp_column else
+                        channel.time_track(absolute_time=True, accuracy="ns")
+                    )
+                    for channel_name, channel in channels_to_export.items()
+                })
+            df = group_df if not df else pd.merge(left=df, right=group_df, left_index=True, right_index=True)
 
         # format for nominal upload
-        time_column = "time_ns"
+        time_column = timestamp_column if timestamp_column else "time_ns"
```

Comment:
there's 5 instances of `if timestamp_column` conditionals in this function. Let's split this into two functions with less conditionals, it's very difficult to grok correctness here